src/main/scala/com/campudus/tableaux/controller/MediaController.scala

```diff
@@ -0,0 +1,135 @@
+package com.campudus.tableaux.controller
+
+import java.util
+import java.util.UUID
+
+import com.campudus.tableaux.TableauxConfig
+import com.campudus.tableaux.database.domain._
+import com.campudus.tableaux.database.model.FolderModel.FolderId
+import com.campudus.tableaux.database.model.{FileModel, FolderModel}
+import com.campudus.tableaux.helper.FutureUtils
+import com.campudus.tableaux.router.UploadAction
+
+import scala.concurrent.{Promise, Future}
+import scala.reflect.io.Path
+import scala.util.Failure
+
+object MediaController {
+  def apply(config: TableauxConfig, folderModel: FolderModel, fileModel: FileModel): MediaController = {
+    new MediaController(config, folderModel, fileModel)
+  }
+}
+
+class MediaController(override val config: TableauxConfig,
+                      override protected val repository: FolderModel,
+                      protected val fileModel: FileModel) extends Controller[FolderModel] {
+
+  import FutureUtils._
+
+  lazy val uploadsDirectory = Path(s"${config.workingDirectory}/${config.uploadsDirectory}")
+
+  def retrieveFolder(id: FolderId): Future[ExtendedFolder] = {
+    for {
+      folder <- repository.retrieve(id)
+      subfolders <- repository.retrieveSubfolders(id)
+      files <- fileModel.retrieveFromFolder(id)
+    } yield ExtendedFolder(folder, subfolders, files)
+  }
+
+  def addNewFolder(name: String, description: String, parent: Option[FolderId]): Future[Folder] = {
+    repository.add(Folder(name, description, parent))
+  }
+
+  def changeFolder(id: FolderId, name: String, description: String, parent: Option[FolderId]): Future[Folder] = {
+    repository.update(Folder(Some(id), name, description, parent, None, None))
+  }
+
+  def deleteFolder(id: FolderId): Future[Folder] = {
+    for {
+      folder <- repository.retrieve(id)
+
+      // delete files
+      _ <- deleteFilesOfFolder(id)
+
+      // delete subfolders
+      _ <- deleteSubfolders(id)
+
+      // delete the folder finally
+      _ <- repository.delete(folder)
+    } yield folder
+  }
+
+  private def deleteSubfolders(id: FolderId): Future[Unit] = {
+    for {
+      folders <- repository.retrieveSubfolders(id)
+      _ <- Future.sequence(folders.map(f => deleteFolder(f.id.get)))
+    } yield ()
+  }
+
+  private def deleteFilesOfFolder(id: FolderId): Future[Unit] = {
+    for {
+      files <- fileModel.retrieveFromFolder(id)
+      _ <- Future.sequence(files.map(f => deleteFile(f.uuid.get)))
+    } yield ()
+  }
+
+  def uploadFile(upload: UploadAction): Future[TemporaryFile] = promisify { p: Promise[TemporaryFile] =>
+    val uuid = UUID.randomUUID()
+    val ext = Path(upload.fileName).extension
+
+    val filePath = uploadsDirectory / Path(s"$uuid.$ext")
+
+    vertx.fileSystem.mkdir(s"$uploadsDirectory", { asyncResult =>
```

Comment:
We could also just create the directory during startup and not try to create it during every upload, WDYT?